src/solvers/refinement/string_constraint_generator.cpp

```diff
@@ -0,0 +1,2147 @@
+
+  return res;
+}
+
+
+string_exprt string_constraint_generatort::add_axioms_for_java_char_array
+(const exprt & char_array)
+{
+  string_exprt res(get_char_type());
+  exprt arr=to_address_of_expr(char_array).object();
+  exprt len=member_exprt(arr, "length", res.length().type());
+  exprt cont=member_exprt(arr, "data", res.content().type());
+  res.op0()=len;
+  res.op1()=cont;
+  return res;
+}
+
+
+string_exprt string_constraint_generatort::add_axioms_for_value_of
+(const function_application_exprt &f)
+{
+  const function_application_exprt::argumentst &args=f.arguments();
+  if(args.size()==3)
+    {
+      string_exprt res(get_char_type());
+      exprt char_array=args[0];
+      exprt offset=args[1];
+      exprt count=args[2];
+      string_exprt str=add_axioms_for_java_char_array(char_array);
+      axioms.push_back(res.has_length(count));
+      symbol_exprt idx=fresh_univ_index("QA_index_value_of");
+      equal_exprt eq(str[plus_exprt(idx, offset)], res[idx]);
+      string_constraintt a1(idx, count, eq);
+      axioms.push_back(a1);
+      return res;
+    }
+  else
+    {
+      assert(args.size()==1);
+      return add_axioms_for_java_char_array(args[0]);
+    }
+}
+
+string_exprt string_constraint_generatort::add_axioms_for_substring
+(const function_application_exprt &f)
+{
+  assert(f.arguments().size()>=2);
+  string_exprt str=add_axioms_for_string_expr(f.arguments()[0]);
+  exprt i(f.arguments()[1]);
+  exprt j;
+  if(f.arguments().size()==3)
+  {
+    j=f.arguments()[2];
+  }
+  else
+  {
+    assert(f.arguments().size()==2);
+    j=str.length();
+  }
+  return add_axioms_for_substring(str, i, j);
+}
+
+string_exprt string_constraint_generatort::add_axioms_for_substring
+(const string_exprt & str, const exprt & start, const exprt & end)
+{
+  symbol_exprt idx=fresh_exist_index("index_substring");
+  assert(start.type()==refined_string_typet::index_type());
+  assert(end.type()==refined_string_typet::index_type());
+  string_exprt res(get_char_type());
+
+  exprt is_empty=res.has_length(refined_string_typet::index_zero());
+  axioms.push_back(implies_exprt(binary_relation_exprt(start, ID_lt, end),
+                                 res.has_length(minus_exprt(end, start))));
+  axioms.push_back(implies_exprt(binary_relation_exprt(start, ID_ge, end),
+                                 is_empty));
+
+  // Warning: check what to do if the string is not long enough
+  axioms.push_back(str.longer(end));
+
+  // forall idx<str.length, str[idx]=arg_str[idx+i]
+  string_constraintt a(idx, res.length(),
+                       equal_exprt(res[idx], str[plus_exprt(start, idx)]));
+  axioms.push_back(a);
+  return res;
+}
+
+string_exprt string_constraint_generatort::add_axioms_for_trim
+(const function_application_exprt &expr)
+{
+  string_exprt str=add_axioms_for_string_expr(args(expr, 1)[0]);
+  string_exprt res(get_char_type());
+  symbol_exprt idx=fresh_exist_index("index_trim");
+  exprt space_char=constant_char(' ');
```

Comment:
You are right, according to the java specification all leading and trailing characters with code below `\u0020` (space char) are removed, I corrected this in https://github.com/diffblue/cbmc/pull/374/commits/57dbbbd1e990cc79bb8fd434ae07285da7d4d386